src/provider/mod.rs

```diff
@@ -112,6 +116,150 @@ impl CustomGetHandler for () {
     }
 }
 
+/// A custom collection parser that allows the user to define what a collection is.
+///
+/// A collection can be anything that contains an ordered sequence of blake3 hashes.
+/// Some collections store links with a fixed size and therefore allow efficient
+/// skipping. Others store links with a variable size and therefore only allow
+/// sequential access.
+///
+/// This API tries to accomodate both use cases. For collections that do not allow
+/// efficient random access, the `skip` method can be implemented by just repeatedly
+/// calling `next`.
+///
+/// For collections that do allow efficient random access, the `skip` method can be
+/// used to move some internal offset.
+pub trait CollectionParser: Send + Default + 'static {
+    /// The iterator that allows to iterate over the collection.
+    type CollectionIter: CollectionIter;
+    /// Future returned by [`CollectionParser::parse`].
+    ///
+    /// For an initial implementation where you don't care that much about allocations,
+    /// this can be just `BoxFuture`. This would be the default if defaults were allowed.
+    type NewFuture: Future<Output = anyhow::Result<Self::CollectionIter>> + Send + 'static;
```

Comment:
I guess the naming oracle was on strike when `NewFuture` was chosen...  Maybe `ParseFuture` is slightly less generic.

Reply:
Lol. The fn was called new at some point. It was more the refactoring consistency fairy that was on strike...

Reply:
still feel like changing this?